### PR TITLE
Add Code Coverage

### LINKED ISSUE #187: Automated Coverage Reports
### Is your feature request related to a problem? Please describe.
Currently our code requires manual code coverage. 

### Describe the solution you'd like
1. It would be nice to automatically check coverage with every pull request
2. Setup a minimum code coverage percent for the project
3. Add coverage badge to README

### Describe alternatives you've considered
[CodeCov GitHub](https://github.com/marketplace/codecov)
[CodeCov Website](https://about.codecov.io/)

[Coverage Badge](https://shields.io/category/coverage)

### Additional context
N/A

### PR DESCRIPTION
**Pull request by JayDwee**
Uses CodeCov to generate coverage reports (free for public repos)
close #187

### CODE REVIEW

**Review by stefan-cooper — APPROVED** [review]
LGTM - very fancy! Way cooler than the jest coverage we've got on frontend but I suppose this is probably another one of those open source is free, pro for private sort of things haha

Does this gate PRs and give a failure if they fall behind a certain coverage threshold %?

**Comment by JayDwee** [comment]
> LGTM - very fancy! Way cooler than the jest coverage we've got on frontend but I suppose this is probably another one of those open source is free, pro for private sort of things haha
> 
> Does this gate PRs and give a failure if they fall behind a certain coverage threshold %?

Currently not, but it can be added with this :https://docs.codecov.com/docs/commit-status 
And we can even make annotations on the files through GitHub with this: https://docs.codecov.com/docs/github-checks

If people want those then I can add them

**Comment by stefan-cooper** [comment]
> Currently not, but it can be added with this :https://docs.codecov.com/docs/commit-status
> And we can even make annotations on the files through GitHub with this: https://docs.codecov.com/docs/github-checks
> 
> If people want those then I can add them

I think it makes sense to add the threshold at the current coverage. That way, when future PRs go in you can be confident that it has been at well covered in tests, and it hasn't decreased the overall coverage. This is what we do in the [frontend](https://github.com/KoalaBotUK/KoalaBotFrontend/blob/9dad7c313adc59d6b1d0ac6a66c42172c945316f/package.json#L99)

**Comment by VirajShah18** [comment]
> > Currently not, but it can be added with this :https://docs.codecov.com/docs/commit-status
> > And we can even make annotations on the files through GitHub with this: https://docs.codecov.com/docs/github-checks
> > If people want those then I can add them
> 
> I think it makes sense to add the threshold at the current coverage. That way, when future PRs go in you can be confident that it has been at well covered in tests, and it hasn't decreased the overall coverage. This is what we do in the [frontend](https://github.com/KoalaBotUK/KoalaBotFrontend/blob/9dad7c313adc59d6b1d0ac6a66c42172c945316f/package.json#L99)

I think that's a good idea to add. Other than that everything looks really good to me!

**Comment by JayDwee** [comment]
Added both, tried to test but didn't work (I'm guessing it doesn't like that master is missing codecov). Ready for another quick review (should be ok)

File: .github/codecov.yml [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+        threshold: 0%
+        base: auto
+        flags:
+          - unit
+#        paths:
+#          - "src"
+        # advanced settings
+        branches:
+          - master
+        if_ci_failed: ignore #success, failure, error, ignore
```
Comment (stefan-cooper) [inline .github/codecov.yml:16]: does this mean it won't fail the job if CI fails?

File: .github/codecov.yml [hunk 2]
```diff
@@ -0,0 +1,18 @@
+coverage:
+  status:
+    project:
+      default:
+        # basic
+        target: auto
+        threshold: 0%
```
Comment (stefan-cooper) [inline .github/codecov.yml:7]: 0% ?

File: .github/codecov.yml (hunk 1 above)
Comment (JayDwee) [reply .github/codecov.yml:16]: Yep, but since the ci is required it shouldnt allow a merge anyway

File: .github/codecov.yml (hunk 2 above)
Comment (JayDwee) [reply .github/codecov.yml:7]: Target is auto (don't drop below head branch aka master) 

threshold is % below the target that is allowed, so 0 meaning you can't lose coverage

**Review by stefan-cooper — APPROVED** [review]

**Pull request merged**